README.md
# Blender Datasmith Export

Export your Blender scene to UE4 using the Datasmith format.

It aims to export all the Datasmith format supports. For now it exports:

* __Meshes__ with normals, vertex colors and up to 8 UV channels.
* __Hierarchy__ is exported keeping meshes references, transforms, parents and
per-instance material overrides from blender.
* __Textures and materials__ are exported using data from the shader graphs.
Materials are closely approximated and a good amount of nodes are supported
(math, mix, fresnel, vertex color and others)
* __Cameras__ are exported trying to match Blender data, keeping focus
distance, focal length, and aperture
* __Lights__ are exported, keeping their type, power, color and size data.
* __Reflection probes__ including Planar, Sphere and Box captures.

Check out an overview of a previous version here:
https://youtu.be/bUUDqerdqAc

## Sample:
You can click the images to open a large preview.

__Blender Eevee:__

<img alt="Blender Eevee render" src="docs/blender.jpg" width="300">

__UE4 using Datasmith:__

<img alt="UE4 render" src="docs/unreal.jpg" width="300">

This result relies on the **DatasmithBlenderContent**, which is a UE4 Plugin
that improves material import compatibility. Consider supporting the project by
purchasing it from [here][gumroad] (Epic Games store support will be added later)

[gumroad]: https://gum.co/DQvTL

This result is in a custom UE4 build, which fixes some issues of the UE4
importer. If you are technical to compile the engine, you can check the fork
[here][ue4 fork]. Hopefully we can get to integrate our fixes into the main
branch.

[ue4 fork]: https://github.com/0xafbf/UnrealEngine/tree/master

## Installation:

Now [__Download the latest development version__][download_link] and install
from the Blender addons preferences pane.

[download_link]: https://github.com/0xafbf/blender-datasmith-export/archive/master.zip

## Frequently Asked Questions:

__Q: Does this support weighted normals/smoothing groups?__

A: Yes, but the plugin is unable to triangulate correctly. For the time
being, you can add a `Triangulate` modifier with the `Keep Normals` option to
work around this.

__Q: Why are some material nodes not exported?__

A: Most of the nodes are exported, but not all of them are imported from the
UE4 side. The Datasmith Blender Additions (mentioned above) improves this by
adding implementations for some of these nodes. There is a
[list of nodes in the wiki] with more information on which nodes are
supported, and which require the UE4 plugin to work.

[list of nodes in the wiki]: https://github.com/0xafbf/blender-datasmith-export/wiki/Supported-Material-Nodes

__Q: What is this "custom build" you talked about earlier?__

A: I modified some of the UE4 build to fix a couple of errors when importing
the scenes generated from Blender. These are related to normal maps import
and very specific import issues with lights. If you're interested you can
check this [custom build discussion].

[custom build discussion]: https://github.com/0xafbf/blender-datasmith-export/issues/25

If you want to support the project, consider supporting via [Patreon].

[patreon]: https://www.patreon.com/0xafbf

Please please, [join the project Discord][join_discord] and share your results!
I want to see what you make and I am open to any feedback you have.

[join_discord]: https://discord.gg/NJt5ADJ



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
Features marked with `*` require the UE4 plugin to work.

## [Unreleased]
### Fixed
* Export metadata is now a flag, off by default, to help in case other addons store info there.
* Libraries with long file paths are now referenced by the blend file name only.

## [1.0.3] 2020-08-19
### Added
+ Texture node now supports box projection.
+ Smoothing groups are now written in geometry data.
+ TGA texture support. _Thanks KeyToon9_

### Fixed
* Vertex normals now export correctly (you may need to use the Triangulate modifier). _Thanks KeyToon9_
* Fixed image files names when first char is unvalid. _Thanks KeyToon9_
* Improved light attenuation radius calculation.
* Animation is now correctly exported for non-root objects.

## [1.0.2] 2020-07-24
### Fixed
* Fixed an issue when writing some custom props, that would break the XML structure.

## [1.0.0] 2020-07-01
### Added
+ Added support for object and meshes metadata (custom properties in Blender).
+ Added an option to prefer custom nodes to generate simpler graphs if you have the UE4 plugin.*
+ Added new material nodes:
  - Texture Coordinates (partial support)*
  - Noise Texture (simulated, not entirely accurate)*
  - UV Map*
  - Geometry (partial support)*
  - Attribute
  - Combine/Separate RGB/XYZ
  - Combine/Separate HSV*
  - Math (some new nodes like SINH, COSH, etc...)*
  - Vector Math
  - Vector Math (Round, Wrap, Project, Reflect)*
  - Mapping (can now do 3d rotations, uses simpler variants when there is no XY rotation)*
  - Checker Texture*

+ **This is a big one:** Object animations are now exported in a Level Sequence.


### Fixed
* Fixed nodegroups with output node named different than "Group Output".
* Names are now sanitized to be only alphanumeric.
* Increased size for datasmith_curves image to allow for more baked curves (up to 1024)
* All UV nodes are upcasted to Vector3 to better reflect Blender operations, and masked down to Vector2 when reading from them.
* Improved images recollection flow. Unpacked files are directly copied and packed files are written from packed data instead of resaving them (this fixes some strange formats like RG8, which Blender can read and pack, but cannot save).
* Improved behavior of **ColorRamp** and **RGBCurveLookup** nodes*
* Improved behavior of RGB sockets connected to VALUE sockets
* Math nodes with `Clamp` option are now wrapped with a `Saturate` operation.
* Normal maps with strength different than 1 are now wrapped with a `FlattenNormal` operation.
* Principled BSDF nodes now write the Specular values.
* UV maps set as `Render Active` are now set at UV0 to be read as default UVs in UE4. \
* Improved handling of some meshes with empty materials or no material slots.
* Tested on **Blender 2.83.1**


## [0.4.0] 2020-03-26
### Added
+ Added support for exporting curves as geometry
+ Added minimal export option, which skips textures for faster export
+ Added new material nodes:
  - Blackbody
  - Bright/Contrast*
  - Object Info
  - Gamma (Power)

### Fixed
* Improved materials rgb_curves node support
* Improved logging output
* We now support **Blender 2.82**

## [0.3.0] 2020-03-09

### Added
+ Added support for multiple uv maps (datasmith supports up to 8)
+ Added support for sphere reflection probes
+ Added support for box reflection probes
+ Added support for planar reflection probes

### Fixed
* Fixed export of material nodetrees inside nodetrees
* Improved export speed of material curves
* Fixed normal map flags for ue4.25
* Fixed profiling flag as export option
* Fixed export for multiple scenes, these scenes export, although some are not
  tested in UE4 yet, tested scenes include:
  + archiviz
  + blender_splash_fishy_cat
  + classroom
  + forest
  + mr_elephant
  + pabellon_barcelona
  + pokedstudio
  + race_spaceship
  + temple
  + the_junk_shop
  + tree_creature
  + wanderer
  + wasp_bot

## [0.2.0]

This was the first release, changelog wasn't used before this.


data_types.py
# Copyright Andrés Botero 2019

import struct
import os
from os import path
import itertools
import bpy
import numpy as np
import logging
import hashlib
log = logging.getLogger("bl_datasmith")

def read_array_data(io, data_struct):
	struct_size = struct.calcsize(data_struct)
	data_struct = "<" + data_struct # force little endianness

	count = struct.unpack("<I", io.read(4))[0]
	data = io.read(count * struct_size)
	unpacked_data = list(struct.iter_unpack(data_struct, data))
	return [tup[0] if len(tup) == 1 else tup for tup in unpacked_data ]


def flatten(it):
	data = []
	for d in it:
		if isinstance(d, float) or isinstance(d, int):
			data.append(d)
		else:
			data += [*d]
	return data

def write_array_data(io, data_struct, data):
	# first get data length
	length = len(data)
	data_struct = '<I' + (data_struct) * length
	flat_data = None
	output = b''
	if isinstance(data, np.ndarray):

		output += struct.pack('<I', length)
		output += data.tobytes()
	else:
		flat_data = flatten(data)
		output = struct.pack(data_struct, length, *flat_data)
	if io:
		io.write(output)
	return output

def write_data(io, data_struct, *args):
	data_struct = '<' + data_struct
	packed = struct.pack(data_struct, *args)
	io.write(packed)

def write_null(io, num_bytes):
	io.write(b'\0' * num_bytes)

def write_string(io, string):
	string_bytes = string.encode('utf-8') + b'\0'
	length = len(string_bytes)
	io.write(struct.pack('<I', length))
	io.write(string_bytes)


def is_char_valid(in_char):
	char_num = ord(in_char)
	if char_num > 47 and char_num < 58:
		return True
	elif char_num > 64 and char_num < 99:
		return True
	elif char_num > 96 and char_num < 123:
		return True
	return False

def sanitize_name(name):
	first_char = ""
	if not is_char_valid(name[0]):
		first_char = "DS"
	sanitized = "".join(letter if is_char_valid(letter) else '_' for letter in name)
	return first_char + sanitized

def f(x):
	return '{:6f}'.format(x)

class Node:
	prefix = ""
	def __init__(self, name, attrs=None, children=None):
		self.name = name
		self.children = children or []
		if attrs:
			assert type(attrs) is dict
		self.attrs = attrs or {}

	def __getitem__(self, key):
		return self.attrs[key]

	def __setitem__(self, key, value):
		self.attrs[key] = value

	def string_rep(self, first=False):
		previous_prefix = Node.prefix
		if first:
			Node.prefix = ""
		else:
			Node.prefix += "\t"
		output = Node.prefix + '<{}'.format(self.name)
		if first:
			Node.prefix = "\n"
		for attr in self.attrs:
			output += ' {key}="{value}"'.format(key=attr, value=self.attrs[attr])

		if self.children:
			output += '>'
			for child in self.children:
				output += str(child)
			if len(self.children) == 1 and type(self.children[0]) == str:
				output += '</{}>'.format(self.name)
			else:
				output += Node.prefix + '</{}>'.format(self.name)
		else:
			output += '/>'
		Node.prefix = previous_prefix
		return output

	def __str__(self):
		return self.string_rep()
	def push(self, value):
		size = len(self.children)
		self.children.append(value)
		return size

def node_value(name, value):
	return Node(name, {'value': f(value)})

class UDMesh():

	def __init__(self, name):
		self.name = name

		self.source_models = 'SourceModels'
		self.struct_property = 'StructProperty'
		self.datasmith_mesh_source_model = 'DatasmithMeshSourceModel'

		self.materials = {}

		self.tris_material_slot = []
		self.tris_smoothing_group = []
		self.vertices = []
		self.triangles = []
		self.vertex_normals = []
		self.uvs = []
		self.vertex_colors = [] # In 0-255 range

		self.test = []

		self.relative_path = None
		self.hash = ''

	# this may need some work, found some documentation:
	# Engine/Source/Developer/Rawmesh
	def write_to_path(self, path):
		with open(path, 'wb') as file:
			log.debug("writing mesh:"+self.name)
			#write_null(file, 8)
			file.write(b'\x01\x00\x00\x00\xfd\x04\x00\x00')

			file_start = file.tell()
			write_string(file, self.name)
			#write_null(file, 5)
			file.write(b'\x00\x01\x00\x00\x00')
			write_string(file, self.source_models)
			write_string(file, self.struct_property)
			write_null(file, 8)

			write_string(file, self.datasmith_mesh_source_model)

			write_null(file, 25)

			size_loc = file.tell() # here we have to write the rawmesh size two times
			write_data(file, 'II', 0, 0) # just some placeholder data, to rewrite at the end

			file.write(b'\x7d\x00\x00\x00\x00\x00\x00\x00') #125 and zero

			#here starts rawmesh
			mesh_start = file.tell()
			file.write(b'\x01\x00\x00\x00') # raw mesh version
			file.write(b'\x00\x00\x00\x00') # raw mesh lic  version

			# further analysis revealed:
			# this loops are per triangle
			write_array_data(file, 'I', self.tris_material_slot)
			write_array_data(file, 'I', self.tris_smoothing_group)


			# per vertex
			write_array_data(file, 'fff', self.vertices) # VertexPositions

			# b2 = write_array_data(file, 'fff', self.test)
			# print(self.vertices)
			# print(self.test)
			# print(b1[0:10])
			# print(b2[0:10])


			# per vertexloop
			write_array_data(file, 'I', self.triangles) # WedgeIndices


			write_null(file, 4) # WedgeTangentX
			write_null(file, 4) # WedgeTangentY
			write_array_data(file, 'fff', self.vertex_normals) # WedgeTangentZ

			num_uvs = len(self.uvs)
			for idx in range(num_uvs):
				write_array_data(file, 'ff', self.uvs[idx]) # WedgeTexCoords[0]

			num_empty_uvs = 8 - num_uvs
			write_null(file, 4 * num_empty_uvs) # WedgeTexCoords[n..7]
			write_array_data(file, 'BBBB', self.vertex_colors) # WedgeColors
			# b2 = write_array_data(file, 'BBBB', self.test) # WedgeTexCoords[0]

			# print("old and new are same? {}".format(b1 == b2))
			# print(b2[4:24])
			# print(self.vertex_colors.tobytes()[:20])
			# print(self.vertex_colors[:20])
			# print(self.test[:20])

			write_null(file, 4) # MaterialIndexToImportIndex

			#here ends rawmesh
			mesh_end = file.tell()

			write_null(file, 16)
			write_null(file, 4)
			file_end = file.tell()

			mesh_size = mesh_end-mesh_start
			file.seek(size_loc)
			write_data(file, 'II', mesh_size, mesh_size)

			file.seek(0)
			write_data(file, 'II', 1, file_end - file_start)

	def node(self):
		n = Node('StaticMesh')
		n['label'] = self.name
		n['name'] = self.name

		for idx, m in self.materials.items():
			n.push(Node('Material', {'id':idx, 'name':m}))
		if self.relative_path:
			path = self.relative_path.replace('\\', '/')
			n.push(Node('file', {'path':path }))
		n.push(Node('LightmapUV', {'value': '-1'}))
		n.push(Node('Hash', {'value': self.hash}))
		return n

	def save(self, basedir, folder_name):
		log.debug("writing mesh:"+self.name)
		self.relative_path = path.join(folder_name, self.name + '.udsmesh')
		abs_path = path.join(basedir, self.relative_path)
		self.write_to_path(abs_path)
		self.hash = calc_hash(abs_path)


def calc_hash(image_path):
	hash_md5 = hashlib.md5()
	with open(image_path, "rb") as f:
		for chunk in iter(lambda: f.read(4096), b""):
			hash_md5.update(chunk)
	return hash_md5.hexdigest()

